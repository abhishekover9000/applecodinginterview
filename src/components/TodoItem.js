import React, { Component } from "react";
import * as constants from "./TodoConst";
import Button from "@material-ui/core/Button";

import Grid from "@material-ui/core/Grid";

class TodoItem extends Component {
  updateItem = action => {
    this.props.update(this.props.index, action);
  };
  render() {
    const { item } = this.props;

    if (item.isDeleteInGrace)
      setTimeout(
        function() {
          this.updateItem(constants.DELETE);
        }.bind(this),
        3000
      );

    return (
      <Grid
        container
        direction="row"
        justify="space-evenly"
        alignItems="center"
        className="item"
      >
        <Grid item>
          {item.isActive ? (
            <b> {item.title} </b>
          ) : (
            <React.Fragment> {item.title} </React.Fragment>
          )}
        </Grid>

        <Grid item>
          {item.isSuccess ? (
            <Button
              variant="text"
              color="secondary"
              onClick={this.updateItem.bind(this, constants.NOT_SUCCESS)}
            >
              {" "}
              Unsuccesfull{" "}
            </Button>
          ) : (
            <Button
              variant="text"
              color="primary"
              onClick={this.updateItem.bind(this, constants.SUCCESS)}
            >
              Successful{" "}
            </Button>
          )}
        </Grid>
        <Grid item>
          {item.isComplete ? (
            <Button
              variant="text outlined"
              color="secondary"
              onClick={this.updateItem.bind(this, constants.INCOMPLETE)}
            >
              {" "}
              Incomplete{" "}
            </Button>
          ) : (
            <Button
              variant="text"
              color="primary"
              onClick={this.updateItem.bind(this, constants.COMPLETE)}
            >
              Complete{" "}
            </Button>
          )}
        </Grid>
        {!item.isComplete && (
          <Grid item>
            {item.isActive ? (
              <Button
                variant="text"
                color="secondary"
                onClick={this.updateItem.bind(this, constants.DEACTIVATE)}
              >
                {" "}
                Deactivate{" "}
              </Button>
            ) : (
              <Button
                variant="text"
                color="primary"
                onClick={this.updateItem.bind(this, constants.ACTIVATE)}
              >
                Activate{" "}
              </Button>
            )}
          </Grid>
        )}
        <Grid item>
          {item.isDeleteInGrace ? (
            <React.Fragment>
              {" "}
              <i>Are you sure? </i>
              <Button
                variant="text"
                color="primary"
                onClick={this.updateItem.bind(this, constants.UNDO)}
              >
                Undo
              </Button>
              <Button
                variant="text"
                color="secondary"
                onClick={this.updateItem.bind(this, constants.DELETE)}
              >
                Delete
              </Button>
            </React.Fragment>
          ) : (
            <Button
              variant="text"
              color="primary"
              onClick={this.updateItem.bind(this, constants.DELETE_GRACE)}
            >
              Delete
            </Button>
          )}
        </Grid>
      </Grid>
    );
  }
}
export default TodoItem;
