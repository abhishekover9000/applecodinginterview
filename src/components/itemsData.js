export const itemsData = [
  {
    name: "test",
    isComplete: false,
    isSuccess: false,
    isDeleted: false,
    isDeleteInGrace: false
  },
  {
    name: "test2",
    isComplete: false,
    isSuccess: true,
    isDeleted: false,
    isDeleteInGrace: false
  }
];
