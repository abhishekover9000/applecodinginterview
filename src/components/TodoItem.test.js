import React from "react";
import { shallow, mount, configure } from "enzyme";
import TodoItem from "./TodoItem";
import { itemsData } from "./itemsData";

describe("TodoItem", () => {
  const mockUpdate = jest.fn();

  let props = {
    item: {
      title: "test",
      isActive: false,
      isComplete: false,
      isSuccess: false,
      isDeleted: false,
      isDeleteInGrace: false
    },
    index: 0,
    update: mockUpdate
  };

  const app = shallow(<TodoItem {...props} />);
  props = {
    item: {
      title: "test",
      isActive: false,
      isComplete: false,
      isSuccess: true,
      isDeleted: false,
      isDeleteInGrace: false
    },
    index: 0,
    update: mockUpdate
  };
  const app2 = shallow(<TodoItem {...props} />);

  it("renders properly", () => {
    expect(app).toMatchSnapshot();
  });

  it("displays success button when isSuccess is false", () => {
    expect(app.find(".successButton").exists()).toEqual(true);
  });

  it("does not display success button when isSuccess is false", () => {
    expect(app2.find(".successButton").exists()).toEqual(false);
  });
});
