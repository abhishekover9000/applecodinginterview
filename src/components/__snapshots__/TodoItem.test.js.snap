// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TodoItem
    index={0}
    item={
      Object {
        "isActive": false,
        "isComplete": false,
        "isDeleteInGrace": false,
        "isDeleted": false,
        "isSuccess": false,
        "title": "test",
      }
    }
    update={
      [MockFunction] {
        "calls": Array [
          Array [
            0,
            "success",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "alignItems": "center",
      "children": Array [
        <WithStyles(Grid)
          item={true}
        >
          <React.Fragment>
             
            test
             
          </React.Fragment>
        </WithStyles(Grid)>,
        <WithStyles(Grid)
          item={true}
        >
          <WithStyles(Button)
            className="successButton"
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Successful
             
          </WithStyles(Button)>
        </WithStyles(Grid)>,
        <WithStyles(Grid)
          item={true}
        >
          <WithStyles(Button)
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Complete
             
          </WithStyles(Button)>
        </WithStyles(Grid)>,
        <WithStyles(Grid)
          item={true}
        >
          <WithStyles(Button)
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Activate
             
          </WithStyles(Button)>
        </WithStyles(Grid)>,
        <WithStyles(Grid)
          item={true}
        >
          <WithStyles(Button)
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Delete
          </WithStyles(Button)>
        </WithStyles(Grid)>,
      ],
      "className": "item",
      "container": true,
      "direction": "row",
      "justify": "space-evenly",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <React.Fragment>
             
            test
             
          </React.Fragment>,
          "item": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              " ",
              "test",
              " ",
            ],
          },
          "ref": null,
          "rendered": Array [
            " ",
            "test",
            " ",
          ],
          "type": Symbol(react.fragment),
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <WithStyles(Button)
            className="successButton"
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Successful
             
          </WithStyles(Button)>,
          "item": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Successful",
              " ",
            ],
            "className": "successButton",
            "color": "primary",
            "onClick": [Function],
            "variant": "text",
          },
          "ref": null,
          "rendered": Array [
            "Successful",
            " ",
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <WithStyles(Button)
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Complete
             
          </WithStyles(Button)>,
          "item": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Complete",
              " ",
            ],
            "color": "primary",
            "onClick": [Function],
            "variant": "text",
          },
          "ref": null,
          "rendered": Array [
            "Complete",
            " ",
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <WithStyles(Button)
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Activate
             
          </WithStyles(Button)>,
          "item": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              "Activate",
              " ",
            ],
            "color": "primary",
            "onClick": [Function],
            "variant": "text",
          },
          "ref": null,
          "rendered": Array [
            "Activate",
            " ",
          ],
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <WithStyles(Button)
            color="primary"
            onClick={[Function]}
            variant="text"
          >
            Delete
          </WithStyles(Button)>,
          "item": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Delete",
            "color": "primary",
            "onClick": [Function],
            "variant": "text",
          },
          "ref": null,
          "rendered": "Delete",
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "alignItems": "center",
        "children": Array [
          <WithStyles(Grid)
            item={true}
          >
            <React.Fragment>
               
              test
               
            </React.Fragment>
          </WithStyles(Grid)>,
          <WithStyles(Grid)
            item={true}
          >
            <WithStyles(Button)
              className="successButton"
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Successful
               
            </WithStyles(Button)>
          </WithStyles(Grid)>,
          <WithStyles(Grid)
            item={true}
          >
            <WithStyles(Button)
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Complete
               
            </WithStyles(Button)>
          </WithStyles(Grid)>,
          <WithStyles(Grid)
            item={true}
          >
            <WithStyles(Button)
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Activate
               
            </WithStyles(Button)>
          </WithStyles(Grid)>,
          <WithStyles(Grid)
            item={true}
          >
            <WithStyles(Button)
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Delete
            </WithStyles(Button)>
          </WithStyles(Grid)>,
        ],
        "className": "item",
        "container": true,
        "direction": "row",
        "justify": "space-evenly",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <React.Fragment>
               
              test
               
            </React.Fragment>,
            "item": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                " ",
                "test",
                " ",
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              "test",
              " ",
            ],
            "type": Symbol(react.fragment),
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(Button)
              className="successButton"
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Successful
               
            </WithStyles(Button)>,
            "item": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "Successful",
                " ",
              ],
              "className": "successButton",
              "color": "primary",
              "onClick": [Function],
              "variant": "text",
            },
            "ref": null,
            "rendered": Array [
              "Successful",
              " ",
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(Button)
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Complete
               
            </WithStyles(Button)>,
            "item": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "Complete",
                " ",
              ],
              "color": "primary",
              "onClick": [Function],
              "variant": "text",
            },
            "ref": null,
            "rendered": Array [
              "Complete",
              " ",
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(Button)
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Activate
               
            </WithStyles(Button)>,
            "item": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                "Activate",
                " ",
              ],
              "color": "primary",
              "onClick": [Function],
              "variant": "text",
            },
            "ref": null,
            "rendered": Array [
              "Activate",
              " ",
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <WithStyles(Button)
              color="primary"
              onClick={[Function]}
              variant="text"
            >
              Delete
            </WithStyles(Button)>,
            "item": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": "Delete",
              "color": "primary",
              "onClick": [Function],
              "variant": "text",
            },
            "ref": null,
            "rendered": "Delete",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
